rust/kernel/of.rs

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Devicetree and Open Firmware abstractions.
+//!
+//! C header: [`include/linux/of_*.h`](../../../../include/linux/of_*.h)
+
+use alloc::boxed::Box;
+
+use crate::{
+    bindings, c_types,
+    error::{Error, Result},
+    types::PointerWrapper,
+    CStr,
+};
+
+use core::mem::transmute;
+
+/// Wraps a kernel Open Firmware / devicetree match table.
+///
+/// Rust drivers may create this structure to match against devices
+/// described in the devicetree.
+///
+/// # Invariants
+///
+/// The last array element is always filled with zeros (the default).
+pub struct OfMatchTable(Box<[bindings::of_device_id; 2]>);
+
+impl OfMatchTable {
+    /// Creates a [`OfMatchTable`] from a single `compatible` string.
+    pub fn new(compatible: &CStr<'static>) -> Result<Self> {
+        let tbl = Box::try_new([
+            Self::new_of_device_id(compatible)?,
+            bindings::of_device_id::default(),
+        ])?;
+        // INVARIANTS: we allocated an array with `default()` as its final
+        // element, therefore that final element will be filled with zeros,
+        // and the invariant above will hold.
+        Ok(Self(tbl))
+    }
+
+    /// Transforms this [`OfMatchTable`] into a raw pointer.
+    ///
+    /// The resulting raw pointer is suitable to be assigned to a
+    /// `bindings::device_driver::of_match_table`.
+    pub(crate) fn into_table_ptr(self) -> *const bindings::of_device_id {
```

Comment:
Are you intentionally consuming the `Box` here?

I think I understand why you'd want to do this, but I'm not a big fan of APIs that leak by design. Ideally we should have a way to recover and free this box if we unregisters the platform device.